lib/gollum/public/gollum/javascript/editor/modes.js.erb

```diff
@@ -0,0 +1,151 @@
+/*  modes.js
+
+    Add highlighting rules for gollum-specific block level syntax to ACE.
+    We first have to extends the Highlighters for each supported syntax,
+    so that later, we can extend the relevant ACE Mode to use the custom Highlighter.
+    We do this for all and only those languages which ACE supports, i.e.,
+    for which it has a specific Highlighter and Mode. So we do not extend e.g. Creole,
+    for which there is no ACE Mode (we use the 'text mode' instead).
+    At the moment there are two gollum-wide block level syntaxes which need to be added:
+
+    * GitHub style code blocks: ``` ... ```
+    * UML blocks: @startuml ... @enduml
+
+    The regexes defined for these below should be equivalent to the regexes used by
+    gollum-lib to extract these blocks.
+
+    ACE Docs:
+      * Highlight Rules: https://cloud9-sdk.readme.io/docs/highlighting-rules
+      * Dynamically generating modes: https://stackoverflow.com/questions/22166784/dynamically-update-syntax-highlighting-mode-rules-for-the-ace-editor
+*/
+
+<%
+codeblock_rule_name = 'GitHubBlock'
+codeblock_rule_id = 'githubblock'
+%>
+
+// The rule defining the github-codeblock state for the ACE Tokenizer
+var <%= codeblock_rule_name %> = [{
+  token: "support.function",
+  regex: /^```$/,
```

Comment:
gollum's current behavior is to only recognize three backticks without any whitespace before/after as a github-style codeblock, so I've copied that behavior here: the regex is strict about whitespace.

However, [here](https://github.com/gollum/gollum-lib/blob/gollum-lib-5.x/lib/gollum-lib/filter/code.rb#L41) it looks like gollum is trying to allow for whitespace (judging by the `([ \t]*)` clauses).

According to [rubular](http:/rubular.com), `\t` is not valid ruby regex. Perhaps this has changed over the years. Question is now: do we want to be lenient about whitespace, or not? If not, should we modify the current regex in gollum-lib?

Reply:
What does github do? I think that should be leading in this case. For me personally I think being strict about whitespace is fine.

Reply:
Github is lenient about the whitespace. That means a bit of tweaking for both the regexes in gollum-lib, and in the JS here.

Reply:
Then again, gollum only supports markdown, and markdown is lenient about whitespace in general. So there is another question about whether it makes sense to be whitespace lenient in all markups.

Reply:
> According to rubular, \t is not valid ruby regex.

This is hard to believe!

Reply:
You're right, `\t` is valid. But for some reason it's not being matched in gollum-lib's regex:
`/^([ \t]*)``` ?([^\r\n]+)?\r?\n(.+?)\r?\n\1```[ \t]*\r?$/m`
Strangely, neither are the spaces -- or any other characters I change the `[ \t]` group to! Any idea why this may be?
(Even more weirdly, removing the `([ \t]*)` group at the beginning, even though it doesn't match anything, will stop a code block from being matched)

Reply:
I've made the regexes in the JS lenient about whitespace, but now noticed the regex in gollum-lib is actually more sophisticated. It allowes whitespace only if the amount of whitespace before the opening backticks is the same as before the final backticks (see the `\1` characters in the regex). I don't see an easy way of implementing this in the JS, as the starting and closing backticks are being detected by different regexes.
Options:
* go full lenient in both gollum and the JS
* be strict in both gollum and the JS
Both options depart from github's behavior. I'll see if there is really no way of doing better in the JS.

Reply:
This would be the new regex for gollum-lib if we go full lenient:
```ruby
/^[ \t]*``` ?([^\r\n]+)?\r?\n(.+?)\r?\n[ \t]*```[ \t]*\r?$/m
```

Reply:
Managed to steal the github code block rule from ace's default markdown highlighter, which seems to cohere with github!

Reply:
GitHub's rules seem simpler than what's implemented by ace: max 3 spaces indent for both the opening and the closing fence. Updated the JS, and the regex in gollum-lib here:
https://github.com/gollum/gollum-lib/pull/338

Reply:
Changed the regexes to support github style fenced block, which means max 3 spaces indent for both the start and end fence. Ace's current implementation is more lenient about whitespace, so to make things consistent I also overrode the default highlighter for markdown (while adding support for `~~~` under markdown only).

https://github.com/gollum/gollum-lib/pull/338 will make gollum-lib consistent with github and the current JS